### PR TITLE
Refs #27859 -- Refactored BaseDatabaseValidation to use check_field_type().

### PR DESCRIPTION
**Pull request by felixxm**
As requested in [comment](https://github.com/django/django/pull/8471/files#r116343787) (see #8471).

### CODE REVIEW

**Comment by timgraham** [comment]
Maybe even less logic could be repeated if a `check_field_type()` hook were added. Most of MySQL's `check_field()` could be moved to the base `check_field()` but it wouldn't run unless the subclass has a `check_field_type()` method. For MySQL it would look like:
```python
def check_field_type(self, field, field_type):
    if (field_type.startswith('varchar') and field.unique and
            (field.max_length is None or int(field.max_length) > 255)):
        errors.append(
            checks.Error(
                'MySQL does not allow unique CharFields to have a max_length > 255.',
                obj=field,
                id='mysql.E001',
            )
        )
```

**Comment by felixxm** [comment]
@timgraham Thanks for review.

**Comment by felixxm** [comment]
I wasn't sure if you wanted to drop `can_check_field_db_type()` or leave it as well.

File: django/db/backends/base/validation.py [hunk 1]
```diff
@@ -7,4 +7,20 @@ def check(self, **kwargs):
         return []
 
     def check_field(self, field, **kwargs):
+        if hasattr(self, 'check_field_type') and self.can_check_field_db_type(field):
+            field_type = field.db_type(self.connection)
+            # Ignore non-concrete fields.
+            if field_type is not None:
+                return self.check_field_type(field, field_type)
         return []
```
Comment (timgraham) [inline django/db/backends/base/validation.py]: I guess this should use the same `errors = [] ... errors.append() ... return errors` approach that similar methods use?

File: django/db/backends/base/validation.py [hunk 2]
```diff
@@ -7,4 +7,17 @@ def check(self, **kwargs):
...
+            db_supports_all_required_features = all(
+                getattr(self.connection.features, feature, False)
+                for feature in field.model._meta.required_db_features
+            )
+            if db_supports_all_required_features:
+                field_type = field.db_type(self.connection)
+                # Ignore non-concrete fields.
+                if field_type is not None:
+                    errors.extend(self.check_field_type(field, field_type))
+        return errors
```
Comment (felixxm) [inline django/db/backends/base/validation.py]: As an alternative we can write:
```python
def check_field(self, field, **kwargs):
    errors = []
    if (hasattr(self, 'check_field_type') and
            # Ignore any related fields.
            not getattr(field, 'remote_field', None) and
            # Ignore fields with unsupported features.
            all(
                getattr(self.connection.features, feature, False)
                for feature in field.model._meta.required_db_features)):
        field_type = field.db_type(self.connection)
        # Ignore non-concrete fields.
        if field_type is not None:
            errors.extend(self.check_field_type(field, field_type))
    return errors
```
I don't have a strong opinion which approach is better.

**Comment by felixxm** [comment]
@timgraham Thanks for review.

File: django/db/backends/base/validation.py [hunk 3]
```diff
@@ -7,4 +7,17 @@ def check(self, **kwargs):
         return []
 
     def check_field(self, field, **kwargs):
-        return []
+        errors = []
+        # Ignore any related fields.
```
Comment (timgraham) [inline django/db/backends/base/validation.py]: I'd just adjust the comments slightly:
```python
# Backends may implement a check_field_type() method.
if hasattr(self, 'check_field_type') and
        # Ignore any related fields.
        not getattr(field, 'remote_field', None):
```

**Pull request merged**